File: code/mobile/lib/utils/mapbox_walk_path_snap.dart
import 'dart:convert';

import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:http/http.dart' as http;
import 'package:pet_trail/utils/geo_utils.dart';

/// Mapbox Map Matching accepts at most 100 coordinates per request.
const int mapboxMatchingMaxInputPoints = 90;

List<LatLng> dedupeCloseConsecutivePoints(
  List<LatLng> points, {
  double minSeparationMeters = 4,
}) {
  if (points.length < 2) return List<LatLng>.from(points);
  final out = <LatLng>[points.first];
  for (var i = 1; i < points.length; i++) {
    final p = points[i];
    final last = out.last;
    final km = haversineDistance(
      last.latitude,
      last.longitude,
      p.latitude,
      p.longitude,
    );
    if (km * 1000 >= minSeparationMeters) {
      out.add(p);
    }
  }
  if (out.length < 2 && points.length >= 2) {
    return [points.first, points.last];
  }
  return out;
}

List<LatLng> subsamplePathToMaxPoints(List<LatLng> points, int maxPoints) {
  final cleaned = dedupeCloseConsecutivePoints(points);
  if (cleaned.length <= maxPoints) return cleaned;
  final out = <LatLng>[];
  for (var i = 0; i < maxPoints; i++) {
    final t = maxPoints == 1 ? 0.0 : i / (maxPoints - 1);
    final idx = (t * (cleaned.length - 1)).round().clamp(0, cleaned.length - 1);
    out.add(cleaned[idx]);
  }
  return dedupeCloseConsecutivePoints(out, minSeparationMeters: 2);
}

/// Returns a dense path snapped to walkable roads, or `null` if the API fails.
Future<List<LatLng>?> fetchMapboxWalkingMatchedPath({
  required List<LatLng> rawPoints,
  required String mapboxAccessToken,
}) async {
  if (mapboxAccessToken.isEmpty || rawPoints.length < 2) return null;
  final sampled = subsamplePathToMaxPoints(rawPoints, mapboxMatchingMaxInputPoints);
  if (sampled.length < 2) return null;

  final coordPath = sampled
      .map((p) => '${p.longitude.toStringAsFixed(6)},${p.latitude.toStringAsFixed(6)}')
      .join(';');
  final encodedPath = Uri.encodeComponent(coordPath);
  final uri = Uri.parse(
    'https://api.mapbox.com/matching/v5/mapbox/walking/$encodedPath'
    '?geometries=geojson&overview=full&steps=false'
    '&access_token=${Uri.encodeQueryComponent(mapboxAccessToken)}',
  );

  try {
    final response = await http.get(uri).timeout(const Duration(seconds: 15));
    if (response.statusCode != 200) return null;
    final body = jsonDecode(response.body) as Map<String, dynamic>;
    final code = body['code'];
    if (code != null && code != 'Ok') return null;
    final matchings = body['matchings'] as List<dynamic>?;
    if (matchings == null || matchings.isEmpty) return null;
    final geometry = matchings.first['geometry'] as Map<String, dynamic>?;
    if (geometry == null || geometry['type'] != 'LineString') return null;
    final coords = geometry['coordinates'] as List<dynamic>?;
    if (coords == null || coords.length < 2) return null;

    final out = <LatLng>[];
    for (final c in coords) {
      if (c is! List || c.length < 2) continue;
      final lng = (c[0] as num).toDouble();
      final lat = (c[1] as num).toDouble();
      out.add(LatLng(lat, lng));
    }
    return out.length >= 2 ? out : null;
  } catch (_) {
    return null;
  }
}

File: code/frontend/src/shared/lib/mapbox-walk-path-snap.ts
export type LatLng = { lat: number; lng: number }

const MAX_INPUT = 90

function haversineKm(a: LatLng, b: LatLng): number {
  const R = 6371
  const dLat = ((b.lat - a.lat) * Math.PI) / 180
  const dLon = ((b.lng - a.lng) * Math.PI) / 180
  const lat1 = (a.lat * Math.PI) / 180
  const lat2 = (b.lat * Math.PI) / 180
  const x =
    Math.sin(dLat / 2) ** 2 +
    Math.sin(dLon / 2) ** 2 * Math.cos(lat1) * Math.cos(lat2)
  return R * (2 * Math.atan2(Math.sqrt(x), Math.sqrt(1 - x)))
}

function dedupeClose(
  points: LatLng[],
  minSeparationMeters: number,
): LatLng[] {
  if (points.length < 2) return [...points]
  const out: LatLng[] = [points[0]]
  for (let i = 1; i < points.length; i++) {
    const p = points[i]
    const last = out[out.length - 1]
    if (haversineKm(last, p) * 1000 >= minSeparationMeters) out.push(p)
  }
  if (out.length < 2 && points.length >= 2) {
    return [points[0], points[points.length - 1]]
  }
  return out
}

function subsampleToMax(points: LatLng[], maxPoints: number): LatLng[] {
  const cleaned = dedupeClose(points, 4)
  if (cleaned.length <= maxPoints) return cleaned
  const out: LatLng[] = []
  for (let i = 0; i < maxPoints; i++) {
    const t = maxPoints === 1 ? 0 : i / (maxPoints - 1)
    const idx = Math.min(
      cleaned.length - 1,
      Math.max(0, Math.round(t * (cleaned.length - 1))),
    )
    out.push(cleaned[idx])
  }
  return dedupeClose(out, 2)
}

/** Geometria alinhada a calçadas/ruas (Mapbox Map Matching), ou `null` se falhar. */
export async function fetchMapboxWalkingMatchedPath(
  rawPoints: LatLng[],
  mapboxAccessToken: string,
): Promise<LatLng[] | null> {
  if (!mapboxAccessToken || rawPoints.length < 2) return null
  const sampled = subsampleToMax(rawPoints, MAX_INPUT)
  if (sampled.length < 2) return null

  const coordPath = sampled
    .map(
      (p) =>
        `${p.lng.toFixed(6)},${p.lat.toFixed(6)}`,
    )
    .join(";")
  const encodedPath = encodeURIComponent(coordPath)
  const token = encodeURIComponent(mapboxAccessToken)
  const url = `https://api.mapbox.com/matching/v5/mapbox/walking/${encodedPath}?geometries=geojson&overview=full&steps=false&access_token=${token}`

  try {
    const ac = new AbortController()
    const t = setTimeout(() => ac.abort(), 20_000)
    const res = await fetch(url, { signal: ac.signal }).finally(() =>
      clearTimeout(t),
    )
    if (!res.ok) return null
    const body = (await res.json()) as {
      code?: string
      matchings?: Array<{ geometry?: { type?: string; coordinates?: unknown } }>
    }
    if (body.code != null && body.code !== "Ok") return null
    const matchings = body.matchings
    if (!matchings?.length) return null
    const geometry = matchings[0]?.geometry
    if (geometry?.type !== "LineString") return null
    const coords = geometry.coordinates
    if (!Array.isArray(coords) || coords.length < 2) return null

    const out: LatLng[] = []
    for (const c of coords) {
      if (!Array.isArray(c) || c.length < 2) continue
      const lng = Number(c[0])
      const lat = Number(c[1])
      if (Number.isFinite(lat) && Number.isFinite(lng)) out.push({ lat, lng })
    }
    return out.length >= 2 ? out : null
  } catch {
    return null
  }
}
